crates/edr_solidity_tests/src/stack_trace.rs

```diff
@@ -0,0 +1,294 @@
+use std::{collections::HashMap, sync::Arc};
+
+use alloy_primitives::{Address, Bytes};
+use edr_solidity::{
+    contract_decoder::{ContractDecoderError, NestedTraceDecoder},
+    exit_code::ExitCode,
+    nested_trace::{
+        CallMessage, CreateMessage, EvmStep, NestedTrace, NestedTraceStep, PrecompileMessage,
+    },
+    solidity_stack_trace::StackTraceEntry,
+    solidity_tracer::{self, SolidityTracerError},
+};
+use foundry_evm::{
+    executors::EvmError,
+    traces::{CallTraceArena, CallTraceStep, TraceKind},
+};
+
+use crate::{
+    constants::{CHEATCODE_ADDRESS, HARDHAT_CONSOLE_ADDRESS},
+    revm::primitives::ruint::aliases::U160,
+};
+
+/// Stack trace generation error during re-execution.
+#[derive(Clone, Debug, thiserror::Error)]
+pub enum StackTraceError {
+    #[error(transparent)]
+    ContractDecoder(#[from] ContractDecoderError),
+    #[error("Unexpected EVM execution error: {0}")]
+    Evm(String),
+    #[error("Invalid root node in call trace arena")]
+    InvalidRootNode,
+    #[error(transparent)]
+    Tracer(#[from] SolidityTracerError),
+}
+
+// `EvmError` is not `Clone`
+impl From<EvmError> for StackTraceError {
+    fn from(value: EvmError) -> Self {
+        Self::Evm(value.to_string())
+    }
+}
+
+/// Compute stack trace based on execution traces.
+/// Assumes last trace is the error one.
+/// Returns `None` if `traces` is empty.
+pub(crate) fn get_stack_trace<NestedTraceDecoderT: NestedTraceDecoder>(
+    contract_decoder: &Arc<NestedTraceDecoderT>,
+    traces: &[(TraceKind, CallTraceArena)],
+) -> Result<Option<Vec<StackTraceEntry>>, StackTraceError> {
+    let mut address_to_creation_code = HashMap::new();
+    let mut address_to_runtime_code = HashMap::new();
+
+    let mut last_trace = None;
+    for (_, trace) in traces {
+        for node in trace.nodes() {
+            let address = node.trace.address;
+            if node.trace.kind.is_any_create() {
+                address_to_creation_code.insert(address, &node.trace.data);
+                address_to_runtime_code.insert(address, &node.trace.output);
+            }
+        }
+        last_trace = Some(trace);
```

Comment:
Wouldn't it be more efficient and easy to read if you do a separate `let last_trace = traces.last()`?

Reply:
Good catch thanks. This is a left-over from an earlier iteration where the function took an `IntoIterator` instead of a slice. Fixed in https://github.com/NomicFoundation/edr/pull/790/commits/6859268e37f49f97d142c846bfb6b94d4ef6c03f